File: app/src/main/java/no/nav/tilbakemeldingsmottak/util/SkjemaUtils.java
package no.nav.tilbakemeldingsmottak.util;

import lombok.NoArgsConstructor;
import no.nav.tilbakemeldingsmottak.rest.serviceklage.domain.Question;

import java.util.Collections;
import java.util.List;
import java.util.Optional;

@NoArgsConstructor
public final class SkjemaUtils {

    public static Optional<Question> getQuestionById(List<Question> questions, String id) {
        return questions.stream()
                .map(question -> id.equals(question.getId())
                        ? Optional.of(question)
                        : Optional.ofNullable(question.getAnswers()).orElse(Collections.emptyList()).stream()
                        .filter(answer -> answer.getQuestions() != null)
                        .map(answer -> getQuestionById(answer.getQuestions(), id))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .findAny())
                .filter(Optional::isPresent)
                .map(Optional::get)
                .findFirst();
    }

}

File: app/src/main/kotlin/no/nav/tilbakemeldingsmottak/util/SkjemaUtils.kt
package no.nav.tilbakemeldingsmottak.util

import no.nav.tilbakemeldingsmottak.model.Question

class SkjemaUtils {

    companion object {
        // OBS: Hvis flere spørsmål har samme id, returneres kun det første
        fun getQuestionById(questions: List<Question>, id: String): Question? {
            return questions.asSequence()
                .mapNotNull { question ->
                    when {
                        id == question.id -> question
                        else -> question.answers?.firstNotNullOfOrNull { answer ->
                            getQuestionById(answer.questions ?: emptyList(), id)
                        }
                    }
                }
                .firstOrNull()
        }

        // Det samme spørsmålet kan være duplisert inn i flere svargrener, så alle forekomster
        // av id-en oppdateres - ikke bare den første.
        fun updateQuestionsById(
            questions: List<Question>?,
            id: String,
            update: (Question) -> Question
        ): List<Question> {
            return questions.orEmpty().map { question ->
                val withUpdatedSubQuestions = question.answers?.let { answers ->
                    question.copy(answers = answers.map { answer ->
                        answer.questions?.let { answer.copy(questions = updateQuestionsById(it, id, update)) }
                            ?: answer
                    })
                } ?: question

                if (withUpdatedSubQuestions.id == id) update(withUpdatedSubQuestions) else withUpdatedSubQuestions
            }
        }
    }


}
